p2p/shim.go

```diff
@@ -0,0 +1,227 @@
+package p2p
+
+import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+var _ Reactor = (*ReactorShim)(nil)
+
+type (
+	CodecShim interface {
+		Marshal(proto.Message) ([]byte, error)
+		Unmarshal([]byte) (proto.Message, error)
+		Validate(proto.Message) error
+	}
+
+	ChannelShim struct {
+		Descriptor *ChannelDescriptor
+		Channel    *Channel
+		InCh       chan Envelope
+		OutCh      chan Envelope
+		PeerErrCh  chan PeerError
+	}
+
+	ReactorShim struct {
+		BaseReactor
+
+		Name         string
+		Codec        CodecShim
+		PeerUpdateCh chan PeerUpdate
+		Channels     map[ChannelID]*ChannelShim
+	}
+)
+
+func NewShim(name string, codec CodecShim, impl Reactor, descriptors []*ChannelDescriptor) *ReactorShim {
+	br := *NewBaseReactor(name, impl)
+
+	channels := make(map[ChannelID]*ChannelShim)
+	for _, cd := range descriptors {
+		cID := ChannelID(cd.ID)
+		inCh := make(chan Envelope)
+		outCh := make(chan Envelope)
+		peerErrCh := make(chan PeerError)
+
+		channels[cID] = &ChannelShim{
+			Descriptor: cd,
+			Channel:    NewChannel(cID, nil, inCh, outCh, peerErrCh),
+			InCh:       inCh,
+			OutCh:      outCh,
+			PeerErrCh:  peerErrCh,
+		}
+	}
+
+	return &ReactorShim{
+		BaseReactor:  br,
+		Name:         name,
+		Codec:        codec,
+		PeerUpdateCh: make(chan PeerUpdate),
+		Channels:     channels,
+	}
+}
+
+// proxyPeerErrors iterates over each p2p Channel and starts a separate go-routine
+// where we listen for peer errors on the channel's PeerErrCh and send a PeerUpdate
+// on the PeerUpdateCh with a status PeerStatusBanned.
+func (rs *ReactorShim) proxyPeerErrors() {
+	for _, c := range rs.Channels {
+		go func(peerErrCh chan PeerError) {
+			for peerErr := range peerErrCh {
+				select {
+				case rs.PeerUpdateCh <- PeerUpdate{PeerID: peerErr.PeerID, Status: PeerStatusBanned}:
+					rs.Logger.Debug(
+						"sent peer update due to error",
+						"reactor", rs.Name, "peer", peerErr.PeerID.String(), "status", PeerStatusBanned, "err", peerErr.Err,
+					)
+
+				default:
+					rs.Logger.Debug(
+						"dropped peer update due to error",
+						"reactor", rs.Name, "peer", peerErr.PeerID.String(), "status", PeerStatusBanned, "err", peerErr.Err,
+					)
+				}
+			}
+		}(c.PeerErrCh)
+	}
+}
+
+// proxyPeerEnvelopes iterates over each p2p Channel and starts a separate
+// go-routine where we listen for outbound envelopes sent during Receive
```

Comment:
nit: not only sent during Receive, can be sent at any time for any reason (e.g. the state sync syncer will solicit snapshot metadata from peers without having received a message from them first).

Reply:
Correct, I'll update the godoc 👍